### PR TITLE
  [Fix #2809] Support requestMessages in web3.js 

### LINKED ISSUE #2809: Offline inbox:  Support requestMessages in web3.js
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a Clojure developer, I want to use web3.js API so that our API calls are consistent.

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature/Refactor

*Summary*: Mirror shh_requestMessages so we can call it via HTTP-RPC.

#### Expected behavior
Request messages like right now but using updated web3.js.

#### Actual behavior
Using JSON-RPC.

### Solution

- [ ] PR to web3.js https://github.com/status-im/web3.js/issues/4
- [ ] Update web3.js dependency in status-react
- [ ] Replace https://github.com/status-im/status-react/blob/develop/src/status_im/protocol/web3/inbox.cljs#L46 with web3 call

### PR DESCRIPTION
**Pull request by foopang**
fixes #2809

- [x] status-im/web3.js#5 should be merged before this one
- [x] Update web3.js dependency in status-react

status: ready

### CODE REVIEW

**Comment by dmitryn** [comment]
@foopang thanks for contribution! 
I think https://github.com/status-im/web3.js/pull/5 should be merged before this one?
Also, it looks like you missed `Update web3.js dependency in status-react` requirement

**Comment by dmitryn** [comment]
I need to mention this PR based on not-yet-merged branch from https://github.com/status-im/status-react/pull/2821

File: src/status_im/protocol/handlers.cljs
Comment (dmitryn) [inline src/status_im/protocol/handlers.cljs]: `web3` twice in args

File: src/status_im/protocol/handlers.cljs
Comment (dmitryn) [inline src/status_im/protocol/handlers.cljs]: please do not change code parts you didn't intend to change

File: src/status_im/protocol/handlers.cljs
Comment (dmitryn) [inline src/status_im/protocol/handlers.cljs]: same, not relevant to changes you're proposing

File: src/status_im/protocol/handlers.cljs
Comment (dmitryn) [inline src/status_im/protocol/handlers.cljs]: i suggest to format arguments in a column as it's done below in `inbox/request-messages` call

**Review by dmitryn — CHANGES_REQUESTED** [review]

**Comment by foopang** [comment]
@dmitryn Thanks! I've updated accordingly.

File: src/status_im/protocol/web3/inbox.cljs
Comment (dmitryn) [inline src/status_im/protocol/web3/inbox.cljs]: I would still use `(if seq)` here https://github.com/bbatsov/clojure-style-guide#nil-punning

File: src/status_im/protocol/web3/inbox.cljs
Comment (dmitryn) [inline src/status_im/protocol/web3/inbox.cljs:34]: What is the format of `error` and `result` args? 
`handle-response` fn is assumed to call callback fn (error-fn or success-fn) with clj data structure, not JSON (see `js->clj` conversion in `parse-json`)

**Review by dmitryn — CHANGES_REQUESTED** [review]
@foopang thanks. I've added couple of new comments, please take a look.

File: src/status_im/protocol/web3/inbox.cljs
Comment (foopang) [reply src/status_im/protocol/web3/inbox.cljs:34]: @dmitryn `error` could be `js/Error` object when you call from web3.js API. It would be `nil` otherwise. And `result` is just `true` when the calback is success from the API calls `markTrustedPeer` and `requestMessages`.
Not sure if I should do the conversion in `parse-json` when directly calling from web3.js API here.

File: src/status_im/protocol/web3/inbox.cljs
Comment (foopang) [reply src/status_im/protocol/web3/inbox.cljs]: @dmitryn I see. But if `js/Error` is being passed to `error`, which is not seqable in ClojureScript, how should it be handled?

Comment (dmitryn) [reply src/status_im/protocol/web3/inbox.cljs]: @foopang good point on `js/Error` is not ISeqable (which makes `seq` and `empty?` fail). So we could check for `nil?`. The thing is `add-peer` call is kind of legacy method which doesn't conform to web3-like response and could return `{"error": ""}` as a success result. I think `parse-json` could also transform such response to standard web3 response by removing blank "error" attribute). Does it make sense to you?

Comment (dmitryn) [reply src/status_im/protocol/web3/inbox.cljs]: @foopang i think we should have 2 separate response handlers: one for web3 response (mark-trusted-peer and request-messages) and one for legacy responses (add-peer) until we would make add-peer response web3-compliant.

Comment (dmitryn) [reply src/status_im/protocol/web3/inbox.cljs]: @foopang current solution with multi-arity fn looks good to me, the only thing i'd change is transforming `{"error": ""}` to `{:result true}` in `parse-json` (to workaround `add-peer` non-standard response).

Comment (foopang) [reply src/status_im/protocol/web3/inbox.cljs]: @dmitryn I see what you mean. It's fixed in [this commit](https://github.com/status-im/status-react/pull/2838/commits/e3b45e5ab4c532e3c38edb271f266a00a9b2e1ee) now.

File: src/status_im/protocol/web3/inbox.cljs
Comment (dmitryn) [inline src/status_im/protocol/web3/inbox.cljs]: Simplified version would be just `(if err)`

**Review by dmitryn — CHANGES_REQUESTED** [review]
Changes on cljs side seems done to me. Please update web3.js dependency.

**Comment by foopang** [comment]
@dmitryn I think I will need https://github.com/status-im/web3.js/pull/5 merged first before I can update web3.js dependency.

**Comment by dmitryn** [comment]
>@dmitryn I think I will need status-im/web3.js#5 merged first before I can update web3.js dependency.

@foopang it seems like PR to web3.js should be opened against `status-develop` branch https://github.com/status-im/status-react/blob/develop/package.json#L92

**Comment by foopang** [comment]
@dmitryn oh, right! Thanks for the heads up.

**Comment by oskarth** [comment]
@foopang In https://github.com/status-im/web3.js/pull/5 I see only requestMessages but not markTrustedPeer, yet PR description here indicates https://github.com/status-im/status-react/issues/2810 is solved. Is this on purpose?

**Comment by foopang** [comment]
@oskarth yes, `markTrustedPeer` is already there https://github.com/status-im/web3.js/blob/status-develop/lib/web3/methods/shh.js#L66-L70

**Comment by dmitryn** [comment]
@foopang i've just merged https://github.com/status-im/web3.js/pull/5 , you could proceed with updating web3.js dependency

**Comment by oskarth** [comment]
> @oskarth yes, markTrustedPeer is already there https://github.com/status-im/web3.js/blob/status-develop/lib/web3/methods/shh.js#L66-L70

Gotcha, I wonder why this wasn't used before instead of JSON-RPC? Do you know @dmitryn?

**Comment by dmitryn** [comment]
>Gotcha, I wonder why this wasn't used before instead of JSON-RPC? Do you know @dmitryn?

@oskarth It even has your TODO comment to use web3.js. I guess it was just postponed for later refactoring.

**Comment by foopang** [comment]
@dmitryn Update web3.js dependency in `package-lock.json`. But I have some setup problem with node v8.9.3 (npm v5.5.1) that `re-natal deps` would modify and delete some files from `node_modules` causing the app fail to build. I have to therefore switch back to node v6.12.2 (npm v3.10.10) in order to make things work. So the file `package-lock.json` was manually edited not generated by `npm`, I couldn't verify it would work with newer version of nodejs with `package-lock.json` support.

**Comment by dmitryn** [comment]
@foopang that's ok because my `npm install` generates same update in package-lock.json as your manual edit.

**Review by dmitryn — APPROVED** [review]

**Comment by andytudhope** [comment]
Hello team! Unfortunately, it does not seem like referencing a fix for two separate bounty issues in 1 PR is going to work - SOB has only picked up the first one referenced here. Shouldn't be too much of a concern - can always close the other issue, retrieve the bounty manually and send it to @foopang - but please don't merge until I have sorted this with Teemu.

**Comment by oskarth** [comment]
> @oskarth It even has your TODO comment to use web3.js. I guess it was just postponed for later refactoring.

Yep, forgot why though. :morty:

File: src/status_im/protocol/web3/inbox.cljs
Comment (oskarth) [inline src/status_im/protocol/web3/inbox.cljs]: keywords truthy, `(:error res)` enough

**Review by oskarth — APPROVED** [review]
Minor comment but LGTM

**Comment by oskarth** [comment]
@dmitryn @foopang do you think we should wait with putting this in QA TODO until is merged https://github.com/status-im/status-react/pull/2821? So we don't get any rebase regressions or so.

**Comment by dmitryn** [comment]
>@dmitryn @foopang do you think we should wait with putting this in QA TODO until is merged #2821? So we don't get any rebase regressions or so.

@oskarth yes, let's wait for 2821 to be merged. It's already in QA TODO btw.

**Comment by oskarth** [comment]
Bounty, medium

**Comment by oskarth** [comment]
@dmitryn Same as this one https://github.com/status-im/status-react/pull/2821#issuecomment-357423031

**Comment by dmitryn** [comment]
@foopang please rebase.

**Comment by andytudhope** [comment]
Also, OpenBounty is not setup to handle a PR that fixes two separate bounty issues. It has picked up the first fix referenced here, but I think the easiest way to handle the bounty for #2810 is to either:
1) Do a no-op PR referencing it
2) Manually reclaim the bounty and pay it out of band (will be cleaner from commit history perspective)
@oskarth WDYT?

**Comment by foopang** [comment]
@andytudhope Done - I've created a separeate PR #3040 for #2810. Please have this one merged before #3040. Thanks!

**Comment by oskarth** [comment]
@dmitryn I think you have this code freshest in your mind, do you want to go ahead and merge this and the other PR if you are happy with it?

**Pull request merged**
